#475: Improve error message for unsupported named parameters

JKDingwall opened the pull request:
Previously just 'Illegal parameter number' was logged, make this a more helpful message.

```
[25-05-06 20:46:04.2629] Slim::Schema::Storage::throw_exception (122) Error: DBI Exception: DBD::mysql::st bind_param failed: named parameters are unsupported: :album [for Statement "
                                        SELECT contributor_album.role AS role, contributors.name AS name, contributors.id AS id
                                        FROM contributor_album
                                        JOIN contributors ON contributors.id = contributor_album.contributor
                                        WHERE contributor_album.album = :album
                                        AND contributor_album.role IN (5,1)
                                        ORDER BY contributor_album.role, contributors.namesort
                                "]
```

JKDingwall committed:
Improve error message for unsupported named parameters:

Previously just 'Illegal parameter number' was logged, make this a more
helpful message.

```
[25-05-06 20:46:04.2629] Slim::Schema::Storage::throw_exception (122) Error: DBI Exception: DBD::mysql::st bind_param failed: named parameters are unsupported: :album [for Statement "
                                        SELECT contributor_album.role AS role, contributors.name AS name, contributors.id AS id
                                        FROM contributor_album
                                        JOIN contributors ON contributors.id = contributor_album.contributor
                                        WHERE contributor_album.album = :album
                                        AND contributor_album.role IN (5,1)
                                        ORDER BY contributor_album.role, contributors.namesort
                                "]
```

dveeden commented:
This probably need a few doc/pod updates and one or more tests.

dveeden commented:
@JKDingwall 

1. Yes just check the current docs and update if needed. Preferably with some suggestion about what to use instead. If you think no updates are needed than that might be ok.
2. Yes. Probably a new test file to test that this does work as intended. If there is a file that already has something very similar that including this in that file would also be fine.

Also note that MySQL might make it easier to add named attributes in the future:
- https://dev.mysql.com/doc/refman/9.3/en/query-attributes.html
- https://dev.mysql.com/doc/c-api/9.3/en/mysql-bind-param.html

JKDingwall committed:
Add tests for new "named parameters are unsupported" error message.

If a named parameter is passed to bind_param() ensure the expected error
message is reported.

JKDingwall committed:
Document that named parameters are not supported in bind_param()

JKDingwall commented:
@dveeden I've added a new test that the expected error message is reported if `bind_param(':name', 1)` is called.  I've also added a couple of sentences to the documentation that named parameters are not supported.  Please let me know you'd like any changes made, thanks!

This is the output from the test:

```
t/45bindnamedparam_error.t ....................
1..11
ok 1
ok 2
ok 3 - create table dbd_mysql_t45bindnamedparam
ok 4 - insert into dbd_mysql_t45bindnamedparam (null, 1)
ok 5
ok 6 - $rows->[0][1] == 1
ok 7
ok 8
ok 9 -
ok 10
ok 11
Argument ":num" isn't numeric in subroutine entry at t/45bindnamedparam.t line 50.
DBD::mysql::st bind_param failed: named parameters are unsupported: :num at t/45bindnamedparam.t line 50.
ok
```

I had seen the documentation about named attributes but after some research I realised that its actually a bit misleading.  The feature provides a way to annotate a query with, for example, the name of script which invoked the query.

- https://stackoverflow.com/questions/77626696/understanding-mysql-stmt-bind-named-param-mysql-c-api
- https://mysql.wisborg.dk/2021/02/02/mysql-query-attributes/

dveeden commented on t/45bindnamedparam_error.t:
```diff
@@ -0,0 +1,53 @@
+plan tests => 11;
+
+SKIP: {
+    skip 'SET @@auto_increment_offset needs MySQL >= 5.0.2', 2 unless $dbh->{mysql_serverversion} >= 50002;
```
DBD::mysql needs MySQL 8.0 or newer, so this shouldn't be needed. But on the other hand... it doesn't hurt either.

JKDingwall replied:
I copied this from t/40bindparam2.t which I used as a template but I have removed it.

dveeden commented on dbdimp.c:
```diff
@@ -4209,6 +4209,23 @@ int dbd_bind_ph(SV *sth, imp_sth_t *imp_sth, SV *param, SV *value,
+      }
+      else
+      {
+        do_error(sth, JW_ERR_ILLEGAL_PARAM_NUM, "<param> could not be coerced to a C string", NULL);
```
Is there a test for this error?

What if there are multiple params. Does it indicate which one is the issue? Could/should it?

JKDingwall replied:
I have added a test for the `else` condition.

From the test execution:
```
Argument ":num" isn't numeric in subroutine entry at t/45bindnamedparam_error.t line 44.
DBD::mysql::st bind_param failed: named parameters are unsupported: :num at t/45bindnamedparam_error.t line 44.
Argument "\0:num" isn't numeric in subroutine entry at t/45bindnamedparam_error.t line 58.
DBD::mysql::st bind_param failed: <param> could not be coerced to a C string at t/45bindnamedparam_error.t line 58.
```

The name of parameter `:num` is reported twice for the  first test, and once for the second.  I'm not sure if the "isn't numeric in subroutine entry" message comes from `SvIV()` or `looks_like_number()`.  I didn't want to try printing the string for the second case in the DBD-mysql error message in case it had weird escape characters.

dveeden replied:
Thanks for the explanation and the second test.

The `Argument "..." isn't numeric in subroutine entry` warning comes from `SvIV(param)` at `dbdimp.c:4190`, which runs before the `looks_like_number` check — `looks_like_number` itself is silent. The `"in subroutine entry"` suffix is Perl's standard label when the warning fires during an XSUB call. That's why `:num` appears twice in the first test (once from SvIV, once from our `do_error`) and once in the second (SvIV warns, but our `else` branch deliberately omits the value).

Your reasoning for not printing the string in the `else` branch is right — once `paramstring[len] != 0` or `strlen != len`, feeding it to `form("%s", ...)` would either truncate at the embedded NUL or read past the buffer. If you want to surface something useful there, `form("named parameters are unsupported: %.*s", (int)len, paramstring)` is bounded by `len` and safe even with embedded NULs (though non-printables would still come through raw).

If the doubled `:num` in the first test bothers you, you can hoist the `looks_like_number(param)` check above the `SvIV(param)` at line 4190 and short-circuit — then SvIV never runs for the named-parameter case and only our error is emitted. Happy to take it either way; the current behavior is fine with me.

*(This reply was drafted with AI assistance.)*

dveeden replied:
I'm ok with merging this as it is now. Or do you want to do one more iteration on it?

dveeden commented on lib/DBD/mysql.pm:
```diff
@@ -1725,6 +1725,10 @@ parameter value and C<execute> the statement again, with other values
 
+MySQL does not support named place holders in C<bind_param>.  If a
+string is passed to C<bind_param> as the parameter index then a
+"named parameters are unsupported" error is reported.
```
```suggestion
`JW_ERR_ILLEGAL_PARAM_NUM` with a message of "named parameters are unsupported" error is reported.
```

btw. I think `JW` stands for Jim Winstead. See `git log`

JKDingwall replied:
Thank you for the suggestion, applied.
